### PR TITLE
use hashcode for service entry store instances

### PR DESCRIPTION
**Pull request by ramaraochavali**
As identified in #25532 , use of config and instance keys in map can occupy huge memory for large service entry deployments. While this is not a o(n^2) issue as was the case with sidecar, this will help us future to scale large deployments.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go [hunk 1]
```diff
@@ -39,6 +41,14 @@ type instancesKey struct {
 	namespace string
 }
 
+func (key instancesKey) HashCode() uint32 {
+	var result uint32
+	result = 31*result + crc32.ChecksumIEEE([]byte(key.hostname))
+	result = 31*result + crc32.ChecksumIEEE([]byte(key.namespace))
+	result = 31*result + crc32.ChecksumIEEE([]byte(key.namespace))
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:48]: two namespace?

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/serviceentry/servicediscovery.go]: yeah. removed

**Comment by howardjohn** [comment]
Is this really needed? It makes sense in the other case since its n^2 and if there is a collision then it just means some extra pushes, not critical failure. Here, if we have a collision we are going to get potentionally dangerous incorrect routing. Additionally, because its not n^2, I am very skeptical that its an issue - consider if we have 10k ServiceEntries, we already have 10k large objects. The keys will be 32b each, so we have 32*2*10k - 500kb, which is negligible?

**Comment by ramaraochavali** [comment]
That is a good question.  As I mentioned in the PR description. Definitely is not a problem now. I was thinking that it might help for deployment that have huge number of service entries. But If you think it is not worth the optimization- I am totally fine with that. Closing now, we can add when it become a problem.

**Pull request closed without merging**